### DISCUSSION TITLE
Not able to solve the Objective Function for MPC

### DISCUSSION THREAD

**Discussion by Meghspace** [yalmip/YALMIP#1230] [General discussion]
I am working on a problem of satellite formation. I am able to generate profile using YALMIP and Gurobi by minimizing the following function.
objective = objective+norm(x-xd1)+ 0*norm(Q*x,1) + 1*norm(R*u{k},1);
where xd are the final reference states. 
but let us suppose i want to make keep a certain distance between the satellite and the reference pints, lets say 10 meters, the code is not giving any output.
objective = objective+abs(norm(x-xd1)-10)+ 0*norm(Q*x,1) + 1*norm(R*u{k},1);

full code 

clc
clear
n0=0.00115691;
 A=[0 0 0 1 0 0
0 0 0 0 1 0
0 0 0 0 0 1
3*n0^2 0 0 0 2*n0 0
0 0 0 -2*n0 0 0
0 0 -(n0)^2 0 0 0];

B=[0 0 0 
   0 0 0 
   0 0 0 
   1 0 0 
   0 1 0 
![error_mpc](https://user-images.githubusercontent.com/127479230/224254895-91288ba0-ba28-4b3d-bcd8-edae1ad1f694.PNG)

   0 0 1]/1000;
C=eye(6);
D=0;
Sys1=ss(A,B,C,D);
Dis_mod=c2d(Sys1,0.1);
Ad=Dis_mod.A;
Bd=Dis_mod.B;
Cd=Dis_mod.C;
Dd=Dis_mod.D;
rank(ctrb(Ad,Bd));
%%
x01=[-70, 0,0, 0 0 0]'/1;

x02=[-50, 0,0, 0 0 0]'/1;

x03=[-30,0,0, 0 0 0]'/1;

xd1=[-30,-20,0,0,0,0]'/1;

xd2=[30,-25,0,0,0,0]'/1;

xd3=[0,32,0,0,0,0]'/1;

%% MPC Parameters

nx = 6; % Number of states
nu = 3; % Number of inputs

% MPC data
Q = eye(6);
R = 1;
N = 2;

%%

x_m(:,1)=x01;

for j=1:500
x = x_m(:,j);
u = sdpvar(repmat(nu,1,N),repmat(1,1,N));
constraints = [];
objective = 0;
for k = 1:N
 x = Ad*x + Bd*u{k};
 objective = objective+norm(x-xd1)+ 0*norm(Q*x,1) + 1*norm(R*u{k},1);
 constraints = [constraints, -100 <= u{k}<= 100];
end
%opts=sdpsettings('solver','gurobi','OutputFlag',0);
%opts = sdpsettings('solver','sdpa');
%optimize(constraints,objective,opts);
optimize(constraints,objective);
U_IN(:,j)=value(u{1});
x_m(:,j+1) = Ad*x_m(:,j) + Bd*U_IN(:,j);
clear constraints;
clear u;
end
figure(1)
plot(x_m(1,:),x_m(2,:))
%%

**Comment by johanlofberg** [comment]
Adding that term to the objective makes the problem brutaly much harder, as it leads to a nonconvex problem. Not only is that a massive issue, the code should not even work, as you use the SOCP-represented norm operator in this nonconvex way, so it should simply yield an error message. I cannot test that though as you code does not run (and is completely broken as you haven't placed it in a code block but simply pasted as text) but I guess you haven't checked that it actually solves the problem (you are not looking at any diagnostics in the code) but then kept running the code anyway thus crashing

Had you used a 1-norm or inf-norm, the term would be accepted, and lead to a MILP-representation

**Comment by Meghspace** [comment]
As far the literature regarding the flight formation using MPC is available I am seeing, I am finding the only possible way was to modify the objective function. Since I am new to the forum I am not very aware of the code cell. 
I tried with 1-norm and inf-norm and still no change. Is there any issue with Gurobi or my formulation. When i tried same using stochastic techniques, I am getting results. 

`clc
clear
n0=0.00115691;
 A=[0 0 0 1 0 0
0 0 0 0 1 0
0 0 0 0 0 1
3*n0^2 0 0 0 2*n0 0
0 0 0 -2*n0 0 0
0 0 -(n0)^2 0 0 0];

B=[0 0 0 
   0 0 0 
   0 0 0 
   1 0 0 
   0 1 0 
   0 0 1]/1000;
C=eye(6);
D=0;
Sys1=ss(A,B,C,D);
Dis_mod=c2d(Sys1,0.1);
Ad=Dis_mod.A;
Bd=Dis_mod.B;
Cd=Dis_mod.C;
Dd=Dis_mod.D;
rank(ctrb(Ad,Bd));
%%
x01=[-70, 0,0, 0 0 0]'/1;

x02=[-50, 0,0, 0 0 0]'/1;

x03=[-30,0,0, 0 0 0]'/1;


xd1=[-30,-20,0,0,0,0]'/1;

xd2=[30,-25,0,0,0,0]'/1;

xd3=[0,32,0,0,0,0]'/1;

%% MPC Parameters

nx = 6; % Number of states
nu = 3; % Number of inputs

% MPC data
Q = eye(6);
R = 1;
N = 2;

%%

x_m(:,1)=x01;

for j=1:500
x = x_m(:,j);
u = sdpvar(repmat(nu,1,N),repmat(1,1,N));
constraints = [];
objective = 0;
for k = 1:N
 k
 x = Ad*x + Bd*u{k};
 objective = objective+abs((norm(x-xd1)))+ 0*norm(Q*x,1) + 1*norm(R*u{k},1);
 constraints = [constraints, -100 <= u{k}<= 100];
end

j;
opts=sdpsettings('solver','gurobi');
%opts = sdpsettings('solver','sdpa');
%optimize(constraints,objective,opts);
optimize(constraints,objective,opts);
U_IN(:,j)=value(u{1});
x_m(:,j+1) = Ad*x_m(:,j) + Bd*U_IN(:,j);
clear constraints;
clear u;
end
figure(1)
plot(x_m(1,:),x_m(2,:))
%%


% U = sdpvar (3, 1) ;
% Y = Ad*x+Bd*U;
% F = set(-100 < U < 100);
% Obj=
% sol = solvesdp (F, Y' *Y+u' *U) ;
`

**Reply by johanlofberg** [reply]
As I said, you have to put your code in a codeblock (<>). If you look at the code you pasted it is broken

**Comment by Meghspace** [comment]
< clc
clear
n0=0.00115691;
 A=[0 0 0 1 0 0
0 0 0 0 1 0
0 0 0 0 0 1
3*n0^2 0 0 0 2*n0 0
0 0 0 -2*n0 0 0
0 0 -(n0)^2 0 0 0];

B=[0 0 0 
   0 0 0 
   0 0 0 
   1 0 0 
   0 1 0 
   0 0 1]/1000;
C=eye(6);
D=0;
Sys1=ss(A,B,C,D);
Dis_mod=c2d(Sys1,0.1);
Ad=Dis_mod.A;
Bd=Dis_mod.B;
Cd=Dis_mod.C;
Dd=Dis_mod.D;
rank(ctrb(Ad,Bd));
%%
x01=[-70, 0,0, 0 0 0]'/1;

x02=[-50, 0,0, 0 0 0]'/1;

x03=[-30,0,0, 0 0 0]'/1;


xd1=[-30,-20,0,0,0,0]'/1;

xd2=[30,-25,0,0,0,0]'/1;

xd3=[0,32,0,0,0,0]'/1;

%% MPC Parameters

nx = 6; % Number of states
nu = 3; % Number of inputs

% MPC data
Q = eye(6);
R = 1;
N = 2;
x_m(:,1)=x01;
for j=1:500
j
Y = x_m(:,j);
U = sdpvar (3, 1);

for k = 1:N
 Y = Ad*Y + Bd*U;
 F = set(-100 <= U <= 100);
end
%opts=sdpsettings('solver','gurobi','OutputFlag',0);
%opts = sdpsettings('solver','sdpa');
%optimize(constraints,objective,opts);
solvemp(F, (Y-xd)'*(Y-xd)')
U_IN(:,j)=U;
x_m(:,j+1) = Ad*x_m(:,j) + Bd*U_IN(:,j);
clear constraints;
clear U;
end
figure(1)
plot(x_m(1,:),x_m(2,:))

>

**Reply by johanlofberg** [reply]
You have to use the code block icon shown in the editor

**Comment by Meghspace** [comment]
`clc
clear
n0=0.00115691;
 A=[0 0 0 1 0 0
0 0 0 0 1 0
0 0 0 0 0 1
3*n0^2 0 0 0 2*n0 0
0 0 0 -2*n0 0 0
0 0 -(n0)^2 0 0 0];

B=[0 0 0 
   0 0 0 
   0 0 0 
   1 0 0 
   0 1 0 
   0 0 1]/1000;
C=eye(6);
D=0;
Sys1=ss(A,B,C,D);
Dis_mod=c2d(Sys1,0.1);
Ad=Dis_mod.A;
Bd=Dis_mod.B;
Cd=Dis_mod.C;
Dd=Dis_mod.D;
rank(ctrb(Ad,Bd));
%%
x01=[-70, 0,0, 0 0 0]'/1;

x02=[-50, 0,0, 0 0 0]'/1;

x03=[-30,0,0, 0 0 0]'/1;


xd1=[-30,-20,0,0,0,0]'/1;

xd2=[30,-25,0,0,0,0]'/1;

xd3=[0,32,0,0,0,0]'/1;

%% MPC Parameters

nx = 6; % Number of states
nu = 3; % Number of inputs

% MPC data
Q = eye(6);
R = 1;
N = 2;
x_m(:,1)=x01;
for j=1:500
j
Y = x_m(:,j);
U = sdpvar (3, 1);

for k = 1:N
 Y = Ad*Y + Bd*U;
 F = set(-100 <= U <= 100);
end
%opts=sdpsettings('solver','gurobi','OutputFlag',0);
%opts = sdpsettings('solver','sdpa');
%optimize(constraints,objective,opts);
solvemp(F, (Y-xd)'*(Y-xd)')
U_IN(:,j)=U;
x_m(:,j+1) = Ad*x_m(:,j) + Bd*U_IN(:,j);
clear constraints;
clear U;
end
figure(1)
plot(x_m(1,:),x_m(2,:))
`

**Reply by johanlofberg** [reply]
Use the preview before you post to make sure it actually looks like a code block

`x = 5*y`

**Comment by Meghspace** [comment]
`for k = 1:N`
`x = Ad*x + Bd*u{k};`
`objective = objective+abs(norm(x-xd1)-10)+ 0*norm(Q*x,1) + 1*norm(R*u{k},1);`
`constraints = [constraints, -100 <= u{k}<= 100];`
`end`
`opts=sdpsettings('solver','gurobi');`
`optimize(constraints,objective,opts);`

The moment I add, abs or inf norm. The optimizer is not giving any output. The control U is coming out as NAN

**Reply by johanlofberg** [reply]
Copy the complete code so it can run.

As I said, `abs(norm()-c)` is not MISOCP-representable so you have no solver which can solve that.  `abs(norm(,1)-c)` is MILP representable and gurobi can solve that, but you still have to check diagnostics to see that the solver actually found a solution (maybe you are setting up an infeasible problem, or maybe the MILP became so hard that it simply gave up before finding a solution)

**Comment by Meghspace** [comment]
`n0=0.00115691;
 A=[0 0 0 1 0 0
0 0 0 0 1 0
0 0 0 0 0 1
3*n0^2 0 0 0 2*n0 0
0 0 0 -2*n0 0 0
0 0 -(n0)^2 0 0 0];`
`B=[0 0 0 
   0 0 0 
   0 0 0 
   1 0 0 
   0 1 0 
   0 0 1]/1000;
C=eye(6);
D=0;
Sys1=ss(A,B,C,D);
Dis_mod=c2d(Sys1,0.1);
Ad=Dis_mod.A;
Bd=Dis_mod.B;
Cd=Dis_mod.C;
Dd=Dis_mod.D;
rank(ctrb(Ad,Bd));`
`x01=[-70, 0,0, 0 0 0]'/1;
x02=[-50, 0,0, 0 0 0]'/1;
x03=[-30,0,0, 0 0 0]'/1;
xd1=[-30,-20,0,0,0,0]'/1;
xd2=[30,-25,0,0,0,0]'/1;
xd3=[0,32,0,0,0,0]'/1;`
`nx = 6; 
nu = 3; 
Q = eye(6);
R = 1;
N = 2;
`
`x_m(:,1)=x01;
for j=1:100
x = x_m(:,j);
u = sdpvar(repmat(nu,1,N),repmat(1,1,N));
constraints = [];
objective = 0;
for k = 1:N
 x = Ad*x + Bd*u{k};
 objective = objective+1*(norm(x-xd1))+ 0*norm(Q*x,1) + 0*norm(R*u{k},1);
 constraints = [constraints, -100 <= u{k}<= 100];
end
opts=sdpsettings('solver','gurobi');
optimize(constraints,objective,opts);
U_IN(:,j)=value(u{1});
x_m(:,j+1) = Ad*x_m(:,j) + Bd*U_IN(:,j);
clear constraints;
clear u;
end
figure(1)
plot(x_m(1,:),x_m(2,:))`

**Reply by johanlofberg** [reply]
You once again sent code that won't run (and is unformatted)

**Reply by Meghspace** [reply]
Sir, is it possible if I mail you the code.

**Reply by johanlofberg** [reply]
No, I don't see what is so hard. Copy the code to the editor. Mark and press code-block.

**Reply by Meghspace** [reply]
Sir, it's a matlab code. Hence whenever I am trying to keep it asking for formatting.

**Reply by johanlofberg** [reply]
You obviously managed to do it above, as you copied and correctly marked as code, which I copied to my matlab, and now copied back here

```
for k = 1:N
    x = Ad*x + Bd*u{k};
    objective = objective+abs(norm(x-xd1)-10)+ 0*norm(Q*x,1) + 1*norm(R*u{k},1);
    constraints = [constraints, -100 <= u{k}<= 100];
end
opts=sdpsettings('solver','gurobi');
optimize(constraints,objective,opts);

```

**Reply by Meghspace** [reply]
Ok, sir. Got it. Will do and send you right away. Thanks

**Comment by Meghspace** [comment]
@johanlofberg  sir, is it possible if I mail you the code.

**Comment by Meghspace** [comment]
`clc`
`clear`
`yalmip('clear')`
`n0=0.00115691;`
` A=[0 0 0 1 0 0`
`0 0 0 0 1 0`
`0 0 0 0 0 1`
`3*n0^2 0 0 0 2*n0 0`
`0 0 0 -2*n0 0 0`
`0 0 -(n0)^2 0 0 0];`

`B=[0 0 0`
   `0 0 0`
   `0 0 0`
   `1 0 0` 
   `0 1 0` 
   `0 0 1]/1000;`
`C=eye(6);`
`D=0;`
`Sys1=ss(A,B,C,D);`
`Dis_mod=c2d(Sys1,0.1);`
`Ad=Dis_mod.A;`
`Bd=Dis_mod.B;`
`Cd=Dis_mod.C;`
`Dd=Dis_mod.D;`
`rank(ctrb(Ad,Bd));`
'%%'
`x01=[-70, 0,0, 0 0 0]'/1;`

`x02=[-50, 0,0, 0 0 0]'/1;`

`x03=[-30,0,0, 0 0 0]'/1;`


`xd1=[-30,-20,0,0,0,0]'/1;`

`xd2=[30,-25,0,0,0,0]'/1;`

`xd3=[0,32,0,0,0,0]'/1;`

`%% MPC Parameters`

`nx = 6; % Number of states`
`nu = 3; % Number of inputs`

`% MPC data`
`Q = eye(6);`
`R = 1;`
`N = 2;`

`%%`

`x_m(:,1)=x01;`

`for j=1:400`
`x = x_m(:,j);`
`u = sdpvar(repmat(nu,1,N),repmat(1,1,N));`
`constraints = [];`
`objective = 0;`
`for k = 1:N`
 `x = Ad*x + Bd*u{k};`
 `objective = objective+(norm(x-xd1))+ 0*norm(Q*x,1) + 0*norm(R*u{k},1);`
 `constraints = [constraints, -100 <= u{k}<= 100];`
`end`
`opts=sdpsettings('solver','gurobi');%'OutputFlag',0);`
`%opts = sdpsettings('solver','scip')`
`%optimize(constraints,objective,opts);`
`optimize(constraints,objective,opts);`
`U_IN(:,j)=value(u{1});`
`x_m(:,j+1) = Ad*x_m(:,j) + Bd*U_IN(:,j);`
`clear constraints;`
`clear u;`
`end`
`figure(1)`
`plot(x_m(1,:),x_m(2,:))`
`%%`

This is working code. 
When we change the norm to norm 1, the optimizer is not working.

**Reply by johanlofberg** [reply]
I changed the objective term to `norm(x-xd1,1)` and it runs like it should so not sure what you are talking about. Not much happens though, the state just stays at the initial state, as your objective does not merit moving towards the goal state (probably as you have such a short horizon, so any movement is penalized and counteracts the improvement in distance)

Your code is violently inefficient though as you keep redefining new variables and constraints inside the loop. You are adviced to read the MPC examples https://yalmip.github.io/example/standardmpc/

**Reply by Meghspace** [reply]
Ok sir. Thanks. Hopefully will get back with some positive results.

**Reply by Meghspace** [reply]
Hi, 
I could solve the problem. Thanks.
But now I have one more problem, 

nx = 6; % Number of states
nu = 6; % Number of inputs
`x = sdpvar(repmat(nx,1,N+1),repmat(1,1,N+1))`
I want to use only the first three elements of each sPDVAR variable
SOMT THING LIKE THIS,
`objective + norm(Q*[x{1:3,k}],1)`


but it is not happening

**Reply by johanlofberg** [reply]
`x{k}(1:3)`

**Reply by Meghspace** [reply]
Sir, following the basic of Yalmip for implicit. I am trying to convert the code for state space. With Input as 6 States and Control Sate. But i AM NOT ABLE TO UNDERSTAND THE architecture of the 

`clc`
`clear`
`yalmip('clear')`
`n0=0.00115691;`
 ```
A=[0 0 0 1 0 0
0 0 0 0 1 0
0 0 0 0 0 1
3*n0^2 0 0 0 2*n0 0
0 0 0 -2*n0 0 0
0 0 -(n0)^2 0 0 0];
```

```
B=[0 0 0 
   0 0 0 
   0 0 0 
   1 0 0 
   0 1 0 
   0 0 1]/100;
```
`C=eye(6);`
`D=0;`
`Sys1=ss(A,B,C,D);`
`Dis_mod=c2d(Sys1,1);`
`Ad=Dis_mod.A;`
`Bd=Dis_mod.B;`
`Cd=Dis_mod.C;`
`Dd=Dis_mod.D;`
`rank(ctrb(Ad,Bd));`
%%
`x01=[-70, 80,0, 0 0 0]'/1;`

`x02=[-50, 0,0, 0 0 0]'/1;`

`x03=[-30,0,0, 0 0 0]'/1;`

`
xd1=[-30,-20,0,0,0,0]'/1;`

`xd2=[30,-25,0,0,0,0]'/1;`

`xd3=[0,32,0,0,0,0]'/1;`

```
%%
%% MPC Parameters
```

```
nx = 6; % Number of states
nu = 3; % Number of inputs

% MPC data
Q = eye(6);
R = 1;
N = 10;

u = sdpvar(repmat(nu,1,N),repmat(1,1,N));
x = sdpvar(repmat(nx,1,N+1),repmat(1,1,N+1));

constraints = [];
objective = 0;

for k = 1:N
 objective = objective + norm(Q*x{k},1) + norm(R*u{k},1);
 constraints = [constraints, x{k+1} == Ad*x{k} + Bd*u{k}];
 constraints = [constraints, -100 <= u{k}<= 100];
end

ops = sdpsettings('solver','gurobi','verbose',0);

controller = optimizer(constraints, objective,ops,x{1},[u{:}]);
x = x01;
implementedU = [];

for i = 1:150
    
  U = controller{x} 
  
  x = Ad*x + Bd*U(1)

  implementedU = [implementedU;U(1)];
end

```

**Reply by johanlofberg** [reply]
There is no question here.

Also, when you show code, put it in a single code block, it looks horrible at the moment.

**Comment by Meghspace** [comment]
Sir, my System is  
X(t+1)=AdX(t)+BdU
X is size(6,1)
U is size(3,1)
I made the necessary changes in the yalmip code, available on "https://yalmip.github.io/example/standardmpc/". 
But the size of u as output of the optimizer is coming as size(30,3) which should be (10,3), As 10 is my sampling interval.
Kindly, tell me where I am getting wrong. 

The changes I made from the original
%% MPC Parameters
nx = 6; % Number of states, Changed from 2 to 6
nu = 3; % Number of inputs, Changed  from 1 to 1

% MPC data
Q = eye(6); % changed from 2 to 1
R = 1;
N = 10; 

The other changes were my own dynamics matrices.


```
clc
clear
yalmip('clear')
n0=0.00115691;

A=[0 0 0 1 0 0
0 0 0 0 1 0
0 0 0 0 0 1
3*n0^2 0 0 0 2*n0 0
0 0 0 -2*n0 0 0
0 0 -(n0)^2 0 0 0];
B=[0 0 0 
   0 0 0 
   0 0 0 
   1 0 0 
   0 1 0 
   0 0 1]/100;
C=eye(6);
D=0;
Sys1=ss(A,B,C,D);
Dis_mod=c2d(Sys1,1);
Ad=Dis_mod.A;
Bd=Dis_mod.B;
Cd=Dis_mod.C;
Dd=Dis_mod.D;
rank(ctrb(Ad,Bd));
%%
x01=[-70, 80,0, 0 0 0]'/1;

x02=[-50, 0,0, 0 0 0]'/1;

x03=[-30,0,0, 0 0 0]'/1;

 xd1=[-30,-20,0,0,0,0]'/1;

xd2=[30,-25,0,0,0,0]'/1;

xd3=[0,32,0,0,0,0]'/1;

%%
%% MPC Parameters
nx = 6; % Number of states
nu = 3; % Number of inputs

% MPC data
Q = eye(6);
R = 1;
N = 10;

u = sdpvar(repmat(nu,1,N),repmat(1,1,N));
x = sdpvar(repmat(nx,1,N+1),repmat(1,1,N+1));

constraints = [];
objective = 0;

for k = 1:N
 objective = objective + norm(Q*x{k},1) + norm(R*u{k},1);
 constraints = [constraints, x{k+1} == Ad*x{k} + Bd*u{k}];
 constraints = [constraints, -100 <= u{k}<= 100];
end

ops = sdpsettings('solver','gurobi','verbose',0);

controller = optimizer(constraints, objective,ops,x{1},[u{:}]);
x = x01;
implementedU = [];

for i = 1:150
    
  U = controller{x} 
  
  x = Ad*x + Bd*U(1)

  implementedU = [implementedU;U(1)];
end
```

**Reply by johanlofberg** [reply]
No, the output from the controller is 3x10, which is exactly what you have requested

```
>> controller(x01)

ans =

  Columns 1 through 8

     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0

  Columns 9 through 10

     0     0
     0     0
     0     0
```

**Reply by johanlofberg** [reply]
and `x = Ad*x + Bd*U(1)` makes no sense. The control input is 3x1 and will be the first column in the returned solution, as you have asked for all predicted inputs stacked in a matrix by columns.

**Reply by Meghspace** [reply]
Thanks Sir.. It is working.. I really owe you a lot... Thanks.. 
Might require your help soon.

**Reply by Meghspace** [reply]
Sir
I want to tap the value of Objective function after every iteration but not able to find any command in Gurobi.
Can you help me with it?
`  optimize(constraints,objective,opts) `

**Reply by johanlofberg** [reply]
I presume by tapping you mean some kind of callback, and that is not possible via YALMIP

**Reply by johanlofberg** [reply]
Unless you mean you simply want to look at the computed value in *your* iteration. Then you define the objective as an output in the optimizer object

**Reply by Meghspace** [reply]
Sir, 
As you told correctly, I just want to see after every iteration of solving the control problem the value of objective function. I want to tap the value of the Objective Function. So that I can demonstrate as the time progresses the Value of objective function is decreasing. 
So, in the current formulation of MPC using YALMIP we are taking the control history and just using the first value of the control array, which we do in MPC, I was looking if I can write the value of the Objective function as well in some text file?

**Reply by johanlofberg** [reply]
hence you output both inputs and objective, `{[u{:}],objective}`

**Reply by Meghspace** [reply]
Sir, can you pleas tell why using 1 in norm function is giving answer while if I am removing 1 there is no solution?

**Reply by johanlofberg** [reply]
not sure what you mean with no solution. If you remove the 1, you will get a 2-norm, and since you use gurobi it should be capable to solve the problem

As always, when developing code where you haven't resolved all issues and is confident everything works you should never use an optimizer as that adds complications, and if you absolutely must use optimizer you should set verbose to 2 so you see what the solver is doing. optimizer constructs are for speeding up simulations only.

to make matters worse, you are not looking at the data you compute, or the diagnostic flag, and thus miss the fact that you start sending matrices to the optimizer object instead of a vector.  this is coming from the fact that the model has three inputs, but you extract U(1) (and this size confusion is visible already in the code as you define R to be a scalar)
